conjur/init_logic.py

```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+
+"""
+InitLogic module
+
+This module is the business logic for writing configuration information
+to the user's machine as well as fetching certificates from Conjur
+"""
+
+# Builtins
+import logging
+import sys
+import os.path
+from pathlib import Path
+
+# Third party
+import socket
+import yaml
+from OpenSSL import SSL
+from OpenSSL.crypto import FILETYPE_PEM
+
+# pylint: disable=raise-missing-from
+class InitLogic:
+    """
+    InitLogic
+
+    This class holds the business logic for population the
+    conjurrc configuration details needed to connect with Conjur
+    """
+    DEFAULT_CONFIG_FILE = os.path.expanduser(os.path.join('~', '.conjurrc'))
+    DEFAULT_PROMPT_RESPONSE=True
+
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def connect_to_socket(hostname, port):
+        """
+        Method for opening a socket with the Conjur server
+        """
+        try:
+            if port is None:
+                port = 443
+
+            ctx = SSL.Context(method=SSL.TLSv1_2_METHOD)
+            # We set the following to false because we want to fetch the
+            # certificates and not use them to validate yet
+            ctx.check_hostname = False
+            ctx.verify_mode = False
+
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sock = SSL.Connection(context=ctx, socket=sock)
+            sock.connect((hostname, port))
+            sock.do_handshake()
+        except Exception:
+            raise Exception(f"Unable to retrieve certificate from {hostname}:{port}")
+
+        logging.info("TLS Connection established. " \
+                     "Fetching certificate from server...")
+
+        return sock
+
+    # pylint: disable=unused-variable
+    @staticmethod
+    def get_certificate(hostname, port, sock, dump_certificate_func):
+        """
+        Method for connecting to Conjur to fetch the certificate chain
+        """
+        try:
+            chain = sock.get_peer_cert_chain()
+            cert = chain[0]
+            fingerprint = cert.digest("sha1").decode("utf-8")
+            # Get the peer chair and format to make it easier to later write to a file
+            readable_certificate=''
+            for (idx, cert) in enumerate(chain):
+                readable_certificate+=str(dump_certificate_func(FILETYPE_PEM, cert),  "utf-8")
+            logging.info("Certificates were fetched successfully")
+        except Exception:
```

Comment:
Bad practice... Can you check possible exception `dump_certificate` might cause?

Reply:
dump_certificate Raises ValueError will change

Reply:
are we only concerned with dump_certificate exceptions? What about if get_peer_cert throws an exception? Thought exception would be the most common ground for both.